cl/antiquary/beacon_states_collector.go

```diff
@@ -0,0 +1,376 @@
+		return err
+	}
+	if err := antiquateFullUint64List(i.slashingsCollector, roundedSlotToDump, state.RawSlashings(), &commonBuffer, compressor); err != nil {
+		return err
+	}
+
+	if err := i.storeEpochData(&commonBuffer, state); err != nil {
+		return err
+	}
+
+	if state.Version() >= clparams.AltairVersion {
+		// dump inactivity scores
+		if err := antiquateFullUint64List(i.inactivityScoresCollector, slot, state.RawInactivityScores(), &commonBuffer, compressor); err != nil {
+			return err
+		}
+		committeeSlot := i.beaconCfg.RoundSlotToSyncCommitteePeriod(slot)
+		committee := *state.CurrentSyncCommittee()
+		if err := i.currentSyncCommitteeCollector.Collect(base_encoding.Encode64ToBytes4(committeeSlot), libcommon.Copy(committee[:])); err != nil {
+			return err
+		}
+
+		committee = *state.NextSyncCommittee()
+		if err := i.nextSyncCommitteeCollector.Collect(base_encoding.Encode64ToBytes4(committeeSlot), libcommon.Copy(committee[:])); err != nil {
+			return err
+		}
+	}
+
+	var b bytes.Buffer
+	if err := i.storeSlotData(&b, state, nil); err != nil {
+		return err
+	}
+
+	return i.stateEventsCollector.Collect(base_encoding.Encode64ToBytes4(slot), events.CopyBytes())
+}
+
+func (i *beaconStatesCollector) storeEpochData(buffer *bytes.Buffer, st *state.CachingBeaconState) error {
+	buffer.Reset()
+	epochData := state_accessors.EpochDataFromBeaconState(st)
+
+	if err := epochData.WriteTo(buffer); err != nil {
+		return err
+	}
+	roundedSlot := i.beaconCfg.RoundSlotToEpoch(st.Slot())
+	return i.epochDataCollector.Collect(base_encoding.Encode64ToBytes4(roundedSlot), libcommon.Copy(buffer.Bytes()))
+}
+
+func (i *beaconStatesCollector) storeSlotData(buffer *bytes.Buffer, st *state.CachingBeaconState, rewardsCollector *eth2.BlockRewardsCollector) error {
+	buffer.Reset()
+	slotData := state_accessors.SlotDataFromBeaconState(st)
+	if rewardsCollector != nil {
+		slotData.AttestationsRewards = rewardsCollector.Attestations
+		slotData.SyncAggregateRewards = rewardsCollector.SyncAggregate
+		slotData.AttesterSlashings = rewardsCollector.AttesterSlashings
+		slotData.ProposerSlashings = rewardsCollector.ProposerSlashings
+	}
+	if err := slotData.WriteTo(buffer); err != nil {
+		return err
+	}
+	return i.slotDataCollector.Collect(base_encoding.Encode64ToBytes4(st.Slot()), libcommon.Copy(buffer.Bytes()))
+}
+
+func (i *beaconStatesCollector) collectEffectiveBalancesDump(slot uint64, uncompressed []byte, buffer *bytes.Buffer, compressor *zstd.Encoder) error {
+	buffer.Reset()
+	compressor.Reset(buffer)
+
+	validatorSetSize := 121
+	for i := 0; i < len(uncompressed)/validatorSetSize; i++ {
+		// 80:88
+		if _, err := compressor.Write(uncompressed[i*validatorSetSize+80 : i*validatorSetSize+88]); err != nil {
+			return err
+		}
+	}
+
+	if err := compressor.Close(); err != nil {
+		return err
+	}
+	roundedSlot := slot - (slot % clparams.SlotsPerDump)
```

Comment:
a bit curious how the underlying data structure was designed.
so we use a key-value storage to keep these data, but here some values share the same key? or I misjudged?